File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-relationship-coverage.py
#!/usr/bin/env python3
"""Fail when Tableau relationship coverage is missing, stale, unwired, or partial."""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent / "lib"))
import relationship_coverage  # noqa: E402


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--workdir", required=True)
    parser.add_argument("--metadata")
    parser.add_argument("--source")
    args = parser.parse_args()
    workdir = Path(args.workdir).expanduser().resolve()
    metadata = (
        Path(args.metadata).expanduser().resolve()
        if args.metadata
        else workdir / "conv-meta.json"
    )
    source = (
        Path(args.source).expanduser().resolve()
        if args.source
        else workdir / "workbook-content.twb"
    )
    artifact = workdir / "relationship-coverage.json"
    model = next(
        (
            path
            for path in (
                workdir / "datamodel-readback.json",
                workdir / "dm-remapped.json",
                workdir / "dm-raw.json",
                workdir / "dm-spec.json",
            )
            if path.is_file()
        ),
        None,
    )

    if not artifact.is_file():
        source_has_graph = (
            source.is_file()
            and relationship_coverage.OBJECT_GRAPH_RE.search(
                source.read_text(encoding="utf-8-sig")
            )
        )
        if not source_has_graph:
            print("relationship coverage: N/A — source has no object-graph")
            return 0
        print(
            f"RELATIONSHIP COVERAGE FAIL: {artifact} is missing for an object-graph source",
            file=sys.stderr,
        )
        return 3
    if not metadata.is_file():
        source_has_graph = (
            source.is_file()
            and relationship_coverage.OBJECT_GRAPH_RE.search(
                source.read_text(encoding="utf-8-sig")
            )
        )
        if source_has_graph:
            print(
                "RELATIONSHIP COVERAGE FAIL: source has an object-graph but "
                f"{metadata} is missing",
                file=sys.stderr,
            )
            return 3
        print("relationship coverage: N/A — no converter metadata and no source object-graph")
        return 0
    try:
        expected = relationship_coverage.from_files(metadata, source, model)
        actual = json.loads(artifact.read_text(encoding="utf-8-sig"))
    except (OSError, ValueError, json.JSONDecodeError) as exc:
        print(f"RELATIONSHIP COVERAGE FAIL: {exc}", file=sys.stderr)
        return 3
    if actual != expected:
        print(
            "RELATIONSHIP COVERAGE FAIL: relationship-coverage.json is stale or "
            "was edited; re-run emit-relationship-coverage.py",
            file=sys.stderr,
        )
        return 3
    if expected["status"] == "fail":
        print(
            "RELATIONSHIP COVERAGE FAIL: "
            f"{len(expected['blockers'])} incomplete source relationship(s)",
            file=sys.stderr,
        )
        for blocker in expected["blockers"]:
            pair = " ↔ ".join(
                str(value)
                for value in (blocker.get("left"), blocker.get("right"))
                if value is not None
            )
            print(
                f"  - {pair or blocker['kind']}: {blocker['reason']}",
                file=sys.stderr,
            )
        return 3
    print(
        f"relationship coverage: {expected['status'].upper()} — "
        f"{expected['wired']}/{expected['serialized']} source relationship(s) "
        "wired completely"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/assert-relationship-coverage.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'optparse'
require_relative 'lib/relationship_coverage'

opts = {}
OptionParser.new do |parser|
  parser.on('--workdir DIR') { |value| opts[:workdir] = File.expand_path(value) }
  parser.on('--metadata PATH') { |value| opts[:metadata] = File.expand_path(value) }
  parser.on('--source PATH') { |value| opts[:source] = File.expand_path(value) }
end.parse!

workdir = opts[:workdir] || abort('usage: assert-relationship-coverage.rb --workdir DIR')
metadata_path = opts[:metadata] || File.join(workdir, 'conv-meta.json')
source_path = opts[:source] || File.join(workdir, 'workbook-content.twb')
artifact_path = File.join(workdir, 'relationship-coverage.json')
model_path = [
  File.join(workdir, 'dm-readback.json'),
  File.join(workdir, 'dm-spec.json'),
  File.join(workdir, 'dm-raw.json')
].find { |path| File.file?(path) }

unless File.file?(artifact_path)
  source_has_graph = File.file?(source_path) &&
                     File.read(source_path, encoding: 'bom|utf-8')
                         .match?(/<(?:[^<>\s]*\.true\.\.\.)?object-graph[\s>\/]/)
  unless source_has_graph
    puts 'relationship coverage: N/A — source has no object-graph'
    exit 0
  end
  warn "RELATIONSHIP COVERAGE FAIL: #{artifact_path} is missing for an object-graph source"
  exit 3
end
unless File.file?(metadata_path)
  source_has_graph = File.file?(source_path) &&
                     File.read(source_path, encoding: 'bom|utf-8')
                         .match?(/<(?:[^<>\s]*\.true\.\.\.)?object-graph[\s>\/]/)
  if source_has_graph
    warn "RELATIONSHIP COVERAGE FAIL: source has an object-graph but #{metadata_path} is missing"
    exit 3
  end
  puts 'relationship coverage: N/A — no converter metadata and no source object-graph'
  exit 0
end

expected = begin
  RelationshipCoverage.from_files(metadata_path, source_path, model_path)
rescue JSON::ParserError, SystemCallError => e
  warn "RELATIONSHIP COVERAGE FAIL: #{e.message}"
  exit 3
end
actual = begin
  JSON.parse(File.read(artifact_path))
rescue JSON::ParserError, SystemCallError => e
  warn "RELATIONSHIP COVERAGE FAIL: #{artifact_path} is unreadable: #{e.message}"
  exit 3
end

unless actual == expected
  warn 'RELATIONSHIP COVERAGE FAIL: relationship-coverage.json is stale or was edited; ' \
       're-run emit-relationship-coverage.rb against conv-meta.json'
  exit 3
end
if expected['status'] == 'fail'
  warn "RELATIONSHIP COVERAGE FAIL: #{expected['blockers'].size} incomplete source relationship(s)"
  expected['blockers'].each do |blocker|
    pair = [blocker['left'], blocker['right']].compact.join(' ↔ ')
    warn "  - #{pair.empty? ? blocker['kind'] : pair}: #{blocker['reason']}"
  end
  exit 3
end

puts "relationship coverage: #{expected['status'].upcase} — " \
     "#{expected['wired']}/#{expected['serialized']} source relationship(s) wired completely"
exit 0
